torrent.go

```diff
@@ -248,21 +300,73 @@ func (t *Torrent) setInfoBytes(b []byte) error {
 		piece.noPendingWrites.L = &piece.pendingWritesMutex
 		missinggo.CopyExact(piece.Hash[:], hash)
 	}
-	for _, conn := range t.conns {
-		if err := conn.setNumPieces(t.numPieces()); err != nil {
-			log.Printf("closing connection: %s", err)
-			conn.Close()
+	if checks != nil {
+		t.checks = checks
+	} else {
+		t.checks = make([]bool, len(ff))
+		for i := range t.checks {
+			t.checks[i] = true
 		}
 	}
+	return nil
+}
+
+// check torrent file for concistency. done will not be called if torrent
+// stoped.
+func (t *Torrent) check() {
+	if t.cl.activeTorrent(t) {
+		// do not allow run integrity checks on running torrents. this is too time
+		// consuming, better redownload parts then check and download at the same
+		// time.
+		return
+	}
+	if t.checking {
+		return
+	}
+	t.checking = true
 	for i := range t.pieces {
 		t.updatePieceCompletion(i)
-		t.pieces[i].QueuedForHash = true
+		p := t.pieces[i]
+		// make it EverHashed false, same as torrent freshly added. so other init
+		// checks flows in order.
+		p.EverHashed = false
+		p.QueuedForHash = true
 	}
 	go func() {
+		defer func() {
+			t.cl.mu.Lock()
+			t.checking = false
+			t.cl.mu.Unlock()
+		}()
 		for i := range t.pieces {
 			t.verifyPiece(i)
+			if t.closed.IsSet() {
+				return
+			}
 		}
 	}()
+}
+
+// for big torrent we need to save CPU time and do not recheck files on disk
+// bettwen application restarts. loadTorrent allows to restore pieces state
```

Comment:
I mentioned this before, this should be done by augmenting the appropriate storage backend. I recently added this to the file and mmap backends, you could test they have the desired effect on your platform.